compiler/rustc_hir/src/lang_items.rs

```diff
@@ -55,17 +55,22 @@ macro_rules! language_item_table {
     (
         $( $(#[$attr:meta])* $variant:ident, $module:ident :: $name:ident, $method:ident, $target:expr, $generics:expr; )*
     ) => {
+        /// A representation of all the valid lang items in Rust.
+        #[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Encodable, Decodable)]
+        pub enum LangItem {
+            $(
+                #[doc = concat!("The `", stringify!($name), "` lang item.")]
+                $(#[$attr])*
+                $variant,
+            )*
+        }
 
-        rustc_data_structures::enum_from_u32! {
-            /// A representation of all the valid lang items in Rust.
-            #[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Encodable, Decodable)]
-            pub enum LangItem {
-                $(
-                    #[doc = concat!("The `", stringify!($name), "` lang item.")]
-                    ///
-                    $(#[$attr])*
-                    $variant,
-                )*
+        impl LangItem {
+            pub fn from_u32(u: u32) -> Option<LangItem> {
+                $(if u == LangItem::$variant as u32 {
+                    return Some(LangItem::$variant)
+                })*
+                None
```

Comment:
This function is only used in the `LanguageItems::iter()`. And each iteration will traverse the whole `LangItem`, which is not efficient. Maybe use a static array to iterate instead?:
```
impl LangItem {
    pub const fn array() -> [LangItem; std::mem::variant_count::<LangItem>()] {
        [$(LangItem::$variant,)*]
    }
}
```

Reply:
I am preserving the existing behaviour. I doubt this code is perf-sensitive?

Reply:
That's OK

Reply:
On second thoughts, this is a good suggestion. I will do it.